File: day8/java/part_1.java
import java.io.*;
import java.util.*;

public class part_1 {
    public static void main(String[] args) throws IOException {
        List<String> grid = new ArrayList<>();
        try (BufferedReader reader = new BufferedReader(new FileReader("input.txt"))) {
            String line;
            while ((line = reader.readLine()) != null) {
                if (!line.trim().isEmpty()) {
                    grid.add(line);
                }
            }
        }

        int N = grid.size();
        int M = grid.get(0).length();
        Map<Character, List<int[]>> nodes = new HashMap<>();

        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                char c = grid.get(i).charAt(j);
                if (c != '.') {
                    nodes.computeIfAbsent(c, k -> new ArrayList<>()).add(new int[]{i, j});
                }
            }
        }

        Set<String> antinodes = new HashSet<>();

        for (List<int[]> nodeList : nodes.values()) {
            int L = nodeList.size();
            for (int i = 0; i < L; i++) {
                for (int j = 0; j < i; j++) {
                    int[] node1 = nodeList.get(i);
                    int[] node2 = nodeList.get(j);
                    int x1 = node1[1], y1 = node1[0];
                    int x2 = node2[1], y2 = node2[0];
                    int newX = x2 + (x2 - x1);
                    int newY = y2 + (y2 - y1);
                    if (newX >= 0 && newX < M && newY >= 0 && newY < N) {
                        antinodes.add(newY + "," + newX);
                    }
                    x1 = node2[1]; y1 = node2[0];
                    x2 = node1[1]; y2 = node1[0];
                    newX = x2 + (x2 - x1);
                    newY = y2 + (y2 - y1);
                    if (newX >= 0 && newX < M && newY >= 0 && newY < N) {
                        antinodes.add(newY + "," + newX);
                    }
                }
            }
        }

        System.out.println(antinodes.size());
    }
}

File: day8/typescript/part_1.ts
import * as fs from 'fs';

const grid = fs.readFileSync('input.txt', 'utf-8')
    .trim()
    .split('\n')
    .filter(line => line.length > 0);

const N = grid.length;
const M = grid[0].length;
const nodes = new Map<string, Array<[number, number]>>();

for (let i = 0; i < N; i++) {
    for (let j = 0; j < M; j++) {
        if (grid[i][j] !== '.') {
            if (!nodes.has(grid[i][j])) {
                nodes.set(grid[i][j], []);
            }
            nodes.get(grid[i][j])!.push([i, j]);
        }
    }
}

const antinodes = new Set<string>();

for (const nodeList of nodes.values()) {
    const L = nodeList.length;
    for (let i = 0; i < L; i++) {
        for (let j = 0; j < i; j++) {
            const node1 = nodeList[i];
            const node2 = nodeList[j];
            for (const [pr1, pr2] of [[node1, node2], [node2, node1]]) {
                const [y1, x1] = pr1;
                const [y2, x2] = pr2;
                const newX = x2 + (x2 - x1);
                const newY = y2 + (y2 - y1);
                if (newX >= 0 && newX < M && newY >= 0 && newY < N) {
                    antinodes.add(`${newY},${newX}`);
                }
            }
        }
    }
}

console.log(antinodes.size);
